Update CI infrastructure

allentiak opened the pull request:
@LevelX2 @jeffwadsworth @JayDi85 

This allows compiling XMage 1.4.43V0 under OpenJDK 11+
Tested with both currently supported OpenJDK versions: 11 (LTS) and 14.

All projects but one build fine. The only exception is Verify (the last one). I would appreciate help with that.

Output and logs below:

### Output

```
Results :

Failed tests:   verifyCards(mage.verify.VerifyCardDataTest): 24 errors in verify

Tests run: 13, Failures: 1, Errors: 0, Skipped: 2

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Mage Root 1.4.43 ................................... SUCCESS [  0.269 s]
[INFO] Mage Framework 1.4.43 .............................. SUCCESS [ 15.357 s]
[INFO] Mage Common Classes 1.4.43 ......................... SUCCESS [  0.717 s]
[INFO] Mage Sets 1.4.43 ................................... SUCCESS [01:00 min]
[INFO] Mage Server Plugins 1.4.43 ......................... SUCCESS [  0.076 s]
[INFO] Mage Player AI 1.4.43 .............................. SUCCESS [  0.382 s]
[INFO] Mage Deck Constructed 1.4.43 ....................... SUCCESS [  0.412 s]
[INFO] Mage Deck Limited 1.4.43 ........................... SUCCESS [  0.097 s]
[INFO] Mage Game Commander Two Player 1.4.43 .............. SUCCESS [  0.097 s]
[INFO] Mage Game Brawl Two Player 1.4.43 .................. SUCCESS [  0.108 s]
[INFO] Mage Game Commander Free For All 1.4.43 ............ SUCCESS [  0.088 s]
[INFO] Mage Game Brawl Free For All 1.4.43 ................ SUCCESS [  0.085 s]
[INFO] Mage Game Free For All 1.4.43 ...................... SUCCESS [  0.070 s]
[INFO] Mage Game Two Player 1.4.43 ........................ SUCCESS [  0.067 s]
[INFO] Mage Player Human 1.4.43 ........................... SUCCESS [  0.198 s]
[INFO] Mage Player AI Minimax 1.4.43 ...................... SUCCESS [  0.204 s]
[INFO] Mage Player AI.MA 1.4.43 ........................... SUCCESS [  0.255 s]
[INFO] Mage Player AI MCTS 1.4.43 ......................... SUCCESS [  0.235 s]
[INFO] Mage Tournament Booster Draft 1.4.43 ............... SUCCESS [  0.641 s]
[INFO] Mage Tournament Sealed 1.4.43 ...................... SUCCESS [  0.056 s]
[INFO] Mage Tournament Constructed 1.4.43 ................. SUCCESS [  0.081 s]
[INFO] Mage Player AI.DraftBot 1.4.43 ..................... SUCCESS [  0.092 s]
[INFO] Mage Game Tiny Leaders Two Player 1.4.43 ........... SUCCESS [  0.054 s]
[INFO] Mage Game Canadian Highlander Two Player 1.4.43 .... SUCCESS [  0.064 s]
[INFO] Mage Game Penny Dreadful Commander Free For All 1.4.43 SUCCESS [  0.052 s]
[INFO] Mage Game Freeform Commander Free For All 1.4.43 ... SUCCESS [  0.070 s]
[INFO] Mage Game Freeform Unlimited Commander 1.4.43 ...... SUCCESS [  0.057 s]
[INFO] Mage Game Freeform Commander Two Player 1.4.43 ..... SUCCESS [  0.056 s]
[INFO] Mage Game Oathbreaker Free For All 1.4.43 .......... SUCCESS [  0.074 s]
[INFO] Mage Game Oathbreaker Two Player 1.4.43 ............ SUCCESS [  0.063 s]
[INFO] Mage Game Momir Basic Two Player 1.4.43 ............ SUCCESS [  0.069 s]
[INFO] Mage Game Momir Basic Free for All 1.4.43 .......... SUCCESS [  0.076 s]
[INFO] Mage Server 1.4.43 ................................. SUCCESS [  1.882 s]
[INFO] Mage Plugins 1.4.43 ................................ SUCCESS [  0.006 s]
[INFO] Mage Counter Plugin 0.1 ............................ SUCCESS [  0.120 s]
[INFO] Mage Client 1.4.43 ................................. SUCCESS [  6.126 s]
[INFO] Mage Server Console 1.4.43 ......................... SUCCESS [  0.479 s]
[INFO] Mage Tests 1.4.43 .................................. SUCCESS [02:21 min]
[INFO] Mage Verify 1.4.43 ................................. FAILURE [ 33.703 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  04:24 min
[INFO] Finished at: 2020-06-22T16:41:17-03:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.12.4:test (default-test) on project mage-verify: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/allentiak/devel/projects/magic-engines/xmage--java/Mage.Verify/target/surefire-reports for the individual test results.
```
### Logs
[openjdk-11.0.7+10-post-Debian-3--mage.verify.VerifyCardDataTest.txt](https://github.com/magefree/mage/files/4815297/openjdk-11.0.7%2B10-post-Debian-3--mage.verify.VerifyCardDataTest.txt)
[openjdk-11.0.7+10-post-Debian-3--TEST-mage.verify.VerifyCardDataTest.xml.txt](https://github.com/magefree/mage/files/4815298/openjdk-11.0.7%2B10-post-Debian-3--TEST-mage.verify.VerifyCardDataTest.xml.txt)

[openjdk-14.0.1+7-Debian-1--TEST-mage.verify.VerifyCardDataTest.xml.txt](https://github.com/magefree/mage/files/4815300/openjdk-14.0.1%2B7-Debian-1--TEST-mage.verify.VerifyCardDataTest.xml.txt)
[openjdk-14.0.1+7-Debian-1--mage.verify.VerifyCardDataTest.txt](https://github.com/magefree/mage/files/4815299/openjdk-14.0.1%2B7-Debian-1--mage.verify.VerifyCardDataTest.txt)

allentiak commented:
Closes #6458

allentiak commented:
Closes #6458

allentiak commented:
From what I could see in the logs, the errors seem to originate in the use of reflection. Maybe merging this PR might be a good opportunity to fix this?

SpeedProg commented:
I just want to add, that I don't need this on my system with OpenJDK+OpenJFX (gentoo). Maybe it is because of they way it is installed (I use gentoos ebuilds from the main repo, with a modified one for openjfx to enable building webkit)

allentiak commented:
On Tue, 23 Jun 2020 at 10:47, SpeedProg <notifications@github.com> wrote:

> I just want to add, that I don't need this on my system with
> OpenJDK+OpenJFX (gentoo). Maybe it is because of they way it is installed
> (I use gentoos ebuilds from the main repo, with a modified one for openjfx
> to enable building webkit)
>

Are you referring to running the released version, or about building from
sources? (I am referring to the second case.)

allentiak commented:
All the output I referred to corresponds to a `mvn install` on the root directory of the repo.

SpeedProg commented:
Not sure what the building from sources refers to, but my OpenJDK+OpenJFX is built from soruce and building XMage from source works with current master (and like old masters since the Java 11 patches were merged).
I always used mvn package to build it because most if the time i pushed it to my server after, but here are logs one for package and one for install.
[mvn_package_output.txt](https://github.com/magefree/mage/files/4820570/mvn_package_output.txt)
[mvn_install_output.txt](https://github.com/magefree/mage/files/4820608/mvn_install_output.txt)

`mvn -version` shows:mvn -version                                                         ✔  10056  18:21:04
Apache Maven 3.6.2 (40f52333136460af0dc0d7232c0dc0bcf0d9e117; 2019-08-27T17:06:16+02:00)
Maven home: /usr/share/maven-bin-3.6
Java version: 11.0.7, vendor: Gentoo, runtime: /usr/lib64/openjdk-11
Default locale: de_DE, platform encoding: UTF-8
OS name: "linux", version: "5.6.14-gentoo", arch: "amd64", family: "unix"
So I am fairly sure it is not using my 1.8.0 jdk by accident for building. Not sure if I can verify this any other way.
``

allentiak commented:
Hmmm.... Interesting.

Could you please pass me the exact hash for master? I will try again...

allentiak commented:
@SpeedProg I forgot to add: Thank you for your constructive reply! :-)

SpeedProg commented:
```
git rev-parse HEAD
7ce721eced0913e727b4d37ac1cf48c61297d962
```

I mean if this change fixes it for you, it should maybe just be added as a dependency, since apparently openjfx is not really part of a standard JDK anymore so it kind of is a dependency now.
I just wanted to note, that it worked for me without it just in case that information is of any use.

allentiak commented:
Thanks again! I'll try it and let you know.

SpeedProg commented:
Hm I took a look and how debian does the openjfx install in bullseye and apparently it only installs the built .jar files into some libs folder.
While on my gentoo it registers the openjfx with a jmod inside the jdk. So I guess that is why it works on gentoo but not on debian :).

allentiak commented:
On Tue, 23 Jun 2020 at 15:49, SpeedProg <notifications@github.com> wrote:
> I mean if this change fixes it for you, it should maybe just be added as a dependency, since
> apparently openjfx is not really part of a standard JDK anymore so it kind of is a dependency now.

That's precisely the reason for all this :-)
I see now that I wasn't explicit about it...
_(AFAIK, OpenJFX has been officially detached from OpenJDK since before Java 8 was realeased. I did a quick search, and it also seems that [AdoptOpenJDK also got hit with this](https://github.com/AdoptOpenJDK/openjdk-build/issues/577). In fact, the only Java 8 "officilal" build I knew that included OpenJFX was Oracle's.)_


> I just wanted to note, that it worked for me without it just in case that information is of any use.

Sure.
What confused me from what you said is that I have (and have had for some time already) both openjdk and openjfx installed. But Maven (as expected) ignores that and uses its own openjfx version.

If there is one thing we all have learned from this, is that Gentoo somehow bundles openjfx into openjdk, and that Debian does not treat openjfx as part of openjdk :-)
(BTW: I tried `mvn install` with the hash you provided me, and it failed. Which proves this.)

apetresc commented:
Stepping back to look at the big picture a bit, is OpenJFX still worth all the trouble its constantly causing? As far as I can tell, literally the only thing it's used for is displaying release notes as HTML.

a) Is there really no other way to do that in modern Java besides JFX?
b) Is it really necessary? Can we have just some rich-text notes in the app with a link to the full set that the user can open in their own browser? Even Magic Arena does that.

allentiak commented:
> Stepping back to look at the big picture a bit, is OpenJFX still worth all the trouble its constantly causing? As far as I can tell, literally the only thing it's used for is displaying release notes as HTML.

Personally, I think the real problem here is depending on a proprietary Java package (Oracle's).

I do agree in that using JFX solely for displaying release notes as HTML is... a little bit over the top.

> a) Is there really no other way to do that in modern Java besides JFX?

No idea.


> b) Is it really necessary? Can we have just some rich-text notes in the app with a link to the full set that the user can open in their own browser? Even Magic Arena does that.

Not my call. I have repeatedly asked for permissions to do simple things, such as labeling issues, and, did not get many _constructive_ answers. In the few cases where I actually got some, they were written in something that, while aiming at being English, is so incomprehensible a simple conversation becomes impossible.

_(Don't take me wrong: I do value the effort of striving to communicate in a language one does not fully master - I am not a native English speaker myself either. That being said, objectively speaking, within the context of international projects such as this one, it is very difficult to communicate properly if the few counterparts that actually answer your messages do not even minimally understand the project's de facto lingua franca - read: English.)_

We are in 2020; Java 8 (LTS) was released in 2014; Java 11 (LTS), in 2018; the next LTS version of the language is due in 2021.
My only aim with this particular PR is being able to build this project on my computer, on a language version younger than 6 :-)

BTW: to which project were you referring as "Magic Arena".

allentiak commented:
It would be nice to get some feedback about how fix the "Verify" project to build (it is the only one I am unable to build).

Here's the log for a clean compitation of this branch rebased onto current master (543ac206be3a19400815a9d4fd5c4d238fdf1bad).
```
% mvn clean
% mvn install |tee  mvn_install--543ac206be3a19400815a9d4fd5c4d238fdf1bad.log
```

[mvn_install--543ac206be3a19400815a9d4fd5c4d238fdf1bad.log](https://github.com/magefree/mage/files/4886482/mvn_install--543ac206be3a19400815a9d4fd5c4d238fdf1bad.log)

JayDi85 commented:
It's outdated logs from mtgjson issue (someday site broken cards data by outdated backup, so verify tests are failed). Find and delete all mtgjson zip files from project folder and run it again.

allentiak commented:
Thanks for your feedback, @JayDi85 !

I deleted the .zip files from Mage.Verify/, and tried again after rebasing (again) onto current master (now at 1db523518b338af1c591e41841af077f17ac703c).

However, I still get the same error...
Somehow, the verification database cannot be created...

I tried this twice: with and without updating the project dependencies.

I checked mtgjson.com, and it seems they have switched to MTGJSON v5.
Maybe the cause of the error is that the current code works with MTGJSON v4?


@JayDi85 Could you please try to reproduce this?

You just need to do a `mvn clean && mvn install` from the project root folder, under OpenJDK.

allentiak commented:
@SpeedProg  Could you try reproducing the error I describe above, please?

JayDi85 commented:
Current code works with latest mtgjson (v5). Show your error logs here.

allentiak commented:
@JayDi85 Please read attached log.

Please note I am using OpenJDK 11.

```
% java -version
openjdk version "11.0.7" 2020-04-14
OpenJDK Runtime Environment (build 11.0.7+10-post-Debian-3)
OpenJDK 64-Bit Server VM (build 11.0.7+10-post-Debian-3, mixed mode, sharing)
```

To reproduce it, `git checkout compile-with-openjdk && cd Mage.Verify && rm *.zip && cd .. && mvn clean && mvn install`.

[mvn_install--1db523518b338af1c591e41841af077f17ac703c.log](https://github.com/magefree/mage/files/4891550/mvn_install--1db523518b338af1c591e41841af077f17ac703c.log)

JayDi85 commented:
I can see h2database errors. Try to delete all `db` folders (in server, client, tests, verify) and run it again (and report logs on errors).

SpeedProg commented:
@allentiak I checked out your compile-with-openjdk branch and deleted *.json.zip files in Mage.Verify and everything in Mage.Verify/db and afterwards maven package ran fine and reported everything as success
```
[INFO] --- maven-jar-plugin:3.2.0:jar (default-jar) @ mage-verify ---
[INFO] Building jar: /home/speedprog/git/mage/Mage.Verify/target/mage-verify.jar
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] Mage Root 1.4.43 ................................... SUCCESS [  0.003 s]
[INFO] Mage Framework 1.4.43 .............................. SUCCESS [ 26.486 s]
[INFO] Mage Common Classes 1.4.43 ......................... SUCCESS [  0.749 s]
[INFO] Mage Sets 1.4.43 ................................... SUCCESS [01:13 min]
[INFO] Mage Server Plugins 1.4.43 ......................... SUCCESS [  0.000 s]
[INFO] Mage Player AI 1.4.43 .............................. SUCCESS [  0.779 s]
[INFO] Mage Deck Constructed 1.4.43 ....................... SUCCESS [  0.052 s]
[INFO] Mage Deck Limited 1.4.43 ........................... SUCCESS [  0.014 s]
[INFO] Mage Game Commander Two Player 1.4.43 .............. SUCCESS [  0.026 s]
[INFO] Mage Game Brawl Two Player 1.4.43 .................. SUCCESS [  0.024 s]
[INFO] Mage Game Commander Free For All 1.4.43 ............ SUCCESS [  0.016 s]
[INFO] Mage Game Brawl Free For All 1.4.43 ................ SUCCESS [  0.026 s]
[INFO] Mage Game Free For All 1.4.43 ...................... SUCCESS [  0.027 s]
[INFO] Mage Game Two Player 1.4.43 ........................ SUCCESS [  0.016 s]
[INFO] Mage Player Human 1.4.43 ........................... SUCCESS [  0.219 s]
[INFO] Mage Player AI Minimax 1.4.43 ...................... SUCCESS [  0.015 s]
[INFO] Mage Player AI.MA 1.4.43 ........................... SUCCESS [  0.017 s]
[INFO] Mage Player AI MCTS 1.4.43 ......................... SUCCESS [  0.041 s]
[INFO] Mage Tournament Booster Draft 1.4.43 ............... SUCCESS [  0.150 s]
[INFO] Mage Tournament Sealed 1.4.43 ...................... SUCCESS [  0.137 s]
[INFO] Mage Tournament Constructed 1.4.43 ................. SUCCESS [  0.059 s]
[INFO] Mage Player AI.DraftBot 1.4.43 ..................... SUCCESS [  0.015 s]
[INFO] Mage Game Tiny Leaders Two Player 1.4.43 ........... SUCCESS [  0.012 s]
[INFO] Mage Game Canadian Highlander Two Player 1.4.43 .... SUCCESS [  0.020 s]
[INFO] Mage Game Penny Dreadful Commander Free For All 1.4.43 SUCCESS [  0.018 s]
[INFO] Mage Game Freeform Commander Free For All 1.4.43 ... SUCCESS [  0.013 s]
[INFO] Mage Game Freeform Unlimited Commander 1.4.43 ...... SUCCESS [  0.024 s]
[INFO] Mage Game Freeform Commander Two Player 1.4.43 ..... SUCCESS [  0.021 s]
[INFO] Mage Game Oathbreaker Free For All 1.4.43 .......... SUCCESS [  0.018 s]
[INFO] Mage Game Oathbreaker Two Player 1.4.43 ............ SUCCESS [  0.017 s]
[INFO] Mage Game Momir Basic Two Player 1.4.43 ............ SUCCESS [  0.017 s]
[INFO] Mage Game Momir Basic Free for All 1.4.43 .......... SUCCESS [  0.014 s]
[INFO] Mage Server 1.4.43 ................................. SUCCESS [ 17.198 s]
[INFO] Mage Plugins 1.4.43 ................................ SUCCESS [  0.001 s]
[INFO] Mage Counter Plugin 0.1 ............................ SUCCESS [  0.022 s]
[INFO] Mage Client 1.4.43 ................................. SUCCESS [ 27.508 s]
[INFO] Mage Server Console 1.4.43 ......................... SUCCESS [  0.389 s]
[INFO] Mage Tests 1.4.43 .................................. SUCCESS [02:04 min]
[INFO] Mage Verify 1.4.43 ................................. SUCCESS [ 51.952 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  05:24 min
```

allentiak commented:
Thanks for your feedbak, @JayDi85.
Thank you for validating my results, @SpeedProg!


Maybe `mvn clean` should be updated to also delete the .zip files and the /db directory?

[Build log: mvn_install--1db523518b338af1c591e41841af077f17ac703c.log](https://github.com/magefree/mage/files/4892192/mvn_install--1db523518b338af1c591e41841af077f17ac703c.log)

It works now!
This is ready to be merged.

Notes:
-The conflict shown by GitHub is due to an outdated dependency I updated.
-The CI failure is due to CI is configured to work with Oracle's JDK8, and not with OpenJDK11 (see #6779).

JayDi85 commented:
Don't merge that issue until it will be prepared with other openjdk 11 migration.

Also change `.travis.yml` to enable openjdk.

allentiak commented:
@JayDi85 In that case, please add those tasks to the description of #6197. Otherwise, you will the only one to know what those tasks are :-)
Probably pinning #6197 may also help triaging this.

allentiak commented:
> Also confirm that client and server works fine under openjdk 11 (with images download, svg mana icons shown, sound plays, etc).

@JayDi85 I will test it and let you know.

allentiak commented:
All old `org.w3c.dom` classes need to be migrated in order for the build to work.
See #6809.

allentiak commented:
@JayDi85 The failed compilation is a Java 9+ bug. I added a workaround for it.

It compiles fine now.

@SpeedProg Could you please give me a hand with manually testing this?
I cannot execute the client, as it cannot find OpenJFX on my machine (you already know why that is :-)
While I find a way to register OpenJFX libs in my system, could you please try building and running a local release?

You just have to:


1. Prepare the local release
``
mvn clean && cd Mage.Verify && rm -rf db/ && cd .. && mvn -DskipTests install && cd Mage.Server && mvn assembly:assembly && cd ../Mage.Client && mvn assebly:assebly && cd ..
```
2. Unpack the content of
`Mage.Client\target\mage-client.zip` and `Mage.Server\target\mage-server.zip` somewhere

3. Start the Server and Client from those directories, and verify they both work.

JayDi85 commented:
BTW you can build release by perl's command line too:
* go to `..\Utils` folder
* run: `perl build-and-package.pl`
* it creates zip file in project's root folder

allentiak commented:
@JayDi85 Thanks for the comment! I had forgotten about that!

SpeedProg commented:
You seem to have eliminated some dependencies from the dependency tree, that now cause the client and server not to be able to communicate any more (notably trove).
`Caused by: java.lang.ClassNotFoundException: gnu.trove.TObjectHashingStrategy`
Loading pictures/symbols works fine.
Running
`mvn dependency:tree -Dverbose | grep -A10 -B10 trove`
on your branch yields no results, while on upstream master it is trove4j and trove are pulled in.

allentiak commented:
@SpeedProg Thanks for your feedback.

I had to exclude `java-xml` due to a Java bug...
(see https://github.com/magefree/mage/pull/6703/commits/1764f13b96e64a8db8a553a037d65a3eabd72129).

Maybe it was because of that?

allentiak commented:
@JayDi85, @SpeedProg 

I think I finally got it.
- [x] rebased on current master (4b6166e4304ab34bbecc641a5b14d8ff47ebe352),
- [x] all deps checked,
- [x] updated Travis CI support,
- [x] works with both OpenJDK 15 and 11,
- [ ] execute local build (pending - see below)

I currently cannot execute the local build. Most likely, this has to do with the fact that OpenJFX is not in Debian's default Java path - I am writing to Debian developers w.r.t. how to "correctly" implement this.

@SpeedProg While I ask the Debian people which would be the "right" way to add OpenJFX's jars to my path...
Could you please verify the local build works in your machine?
(This is, verifying a single game with both the locally built client and server under java 11 -ideallly, under both 11 and 15- works successfully?)

SpeedProg commented:
I noticed that the card/token image download seem to be broken, idk but when selecting scryfall as source it doesn't says it downloaded everything except 4 images, but also says at the top that there is like 450 cards and about 1k tokens missing. Not sure what this is about and I am pretty sure it has nothing to do with the changes in this. Just mentioning it because it might.

Also when looking at the diffs, I noticed that you upgraded several libraries (or changed dependencies entirely to different libraries), not too sure that is a good idea mixing those in here unless it is needed to get it to run. Also when modifying code you just changed the line endings in those files which makes those changes really hard to read.

In the end it builds fine for me and I can play games against AI with both client and server build and running with openjdk11, not gonna test it for anything higher, since official gentoo has no ebuilds and I have to much to do to build openjdk by hand currently.

JayDi85 commented:
@SpeedProg try to change download threads to 1-2 -- it can help with scryfall IP ban for too much images download.

allentiak commented:
> when looking at the diffs, I noticed that you upgraded several libraries (or changed dependencies entirely to different libraries), not too sure that is a good idea mixing those in here unless it is needed to get it to run.

@SpeedProg OK. I have moved the dependency updates to another branch. I will propose the deps updates in a new PR.

@JayDi85 I have also rebased this PR on current master (8c7d7f0b7732c651d47c9fb0b541881ea8be5d72).
I think this is ready to be merged now.


> In the end it builds fine for me and I can play games against AI with both client and server build and running with openjdk11

Cool!
_(I haven't been able to get a proper answer on [my question on openjfx to Debian devs](https://lists.debian.org/debian-java/2020/10/msg00003.html). I think I will have to open a bug on the package...)_


> not gonna test it for anything higher, since official gentoo has no ebuilds and I have to much to do to build openjdk by hand currently.

No problem. AFAIK, there hasn't been any breaking change since OpenJDK9.

allentiak commented:
@SpeedProg Thank you very much for your help on this!

allentiak commented:
@SpeedProg w.r.t. the dependencies, I have just opened #7128.

allentiak commented:
@JayDi85 Is there anything else I should do in order for this to be merged?

Just to recap:
- [x] rebased on current master (4b6166e4304ab34bbecc641a5b14d8ff47ebe352),
- [x] updated Travis CI support (OpenJDK 11),
- [x] add Travis CI support for future OpenJDK versions (build also against the latest OpenJDK EA[*] version, but without breaking in case of failure)
- [x] local release: build and package
- [X] local release: execute

[*]: EA: early access

JayDi85 commented:
1. Test GUI development under NetBeans (preferred netbeans version is 8) with OpenJDK. You must able to open dialog forms like Preferences, edit it, save and compile without errors (by netbeans or another tool).
2. Test clients under 8, 9, 11 connecting to server under openjdk 11+ (is it able to receive some messages/errors or it's just freeze). 
3. And just wait for the right moment.

allentiak commented:
@JayDi85 Thanks for your answer.

Considering that both items 1. and 2. have already been successfully completed (by me and @SpeedProg)...
I guess it is up to you to decide on nr. 3.? :-)

allentiak commented:
Just rebased against current master (9f7a873e5b).

allentiak commented:
@JayDi85 BTW: NetBeans' latest version is 12.1.
Older versions are currently unsupported. (Specifically, NetBeans 8 is around four years old now.)

allentiak commented:
@JayDi85
JFYTK: I've just successfully verified that the official 1.4.46V0-beta6 (2020-10-28) release runs perfectly fine under OpenJDK 11 and 15.
I have also successfully tested both components (client and server) in both environments for playing local games.

allentiak commented:
Just rebased against current master (fec9da5).

t-my commented:
@JayDi85 could we do something for this?

JayDi85 commented:
@t-my yes, it will be released soon with new launcher (https://github.com/magefree/Launcher/pull/14)

SpeedProg commented:
You can Compiler xmage in openjdk11 without these patches if you habe openjfx jmods properly installed. Wich seems ti happen if you usw the buildscript to install it.

cognivore commented:
I managed to compile it today, but I can't run yet because of JFX, which I'll try to patch out over weekend.

Alex-Vasile commented:
These changes are no longer needed as the javafx dependency has been removed.

However I can't get the current head to compile with openjdk 11 or 18 on my Mac as it produces a different ClassNotFoundException.

```
mage.server.util.ConfigurationException: javax.xml.bind.JAXBException: Implementation of JAXB-API has not been found on module path or classpath.
 - with linked exception:
[java.lang.ClassNotFoundException: com.sun.xml.bind.v2.ContextFactory]

	at mage.server.util.ConfigFactory.loadFromFile(ConfigFactory.java:17)
	at org.mage.test.serverside.base.MageTestPlayerBase.init(MageTestPlayerBase.java:139)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.RunBefores.invokeMethod(RunBefores.java:33)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:42)
	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:80)
	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:72)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:71)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
```

Alex-Vasile commented on Mage.Client/pom.xml:
```diff
@@ -175,6 +175,30 @@
             <artifactId>unbescape</artifactId>
             <version>1.1.6.RELEASE</version>
         </dependency>
+        <dependency>
+            <groupId>org.openjfx</groupId>
+            <artifactId>javafx-controls</artifactId>
+            <version>11.0.2</version>
+            <type>pom</type>
+        </dependency>
+        <dependency>
+            <groupId>org.openjfx</groupId>
+            <artifactId>javafx-fxml</artifactId>
+            <version>11.0.2</version>
+            <type>pom</type>
+        </dependency>
+        <dependency>
+            <groupId>org.openjfx</groupId>
+            <artifactId>javafx-web</artifactId>
+            <version>11.0.2</version>
+            <type>pom</type>
+        </dependency>
+        <dependency>
+            <groupId>org.openjfx</groupId>
+            <artifactId>javafx-swing</artifactId>
+            <version>11.0.2</version>
+            <type>pom</type>
+        </dependency>
```
Can you please remove the javafx dependencies (or would you like me to) as they are no longer needed. The change to the travis script are still good.

allentiak replied:
Done

allentiak commented:
Done. Also renamed PR accordingly.

Thanks for taking this over. I'm happy that, finally, someone with merging power starts giving this some attention.

Alex-Vasile commented:
Will merge once Tarvis CI is back up and running. It seems to have disappeared from the project...
